### PR TITLE
Feature/report (#7)

### PR DESCRIPTION
**Pull request by Aruisehu**
Made report system by centralizing into a Report object.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -162,14 +157,21 @@ def random
...
 
     def report
-        title('Report post')
-        render component 'posts/report'
+        if current_user
+            title('Report post')
+            render component 'posts/report'
+        else
+            flash[:error] = 'You should be connected to report a post'
+            redirect_to root_path
```
Comment (eternialz) [inline app/controllers/posts_controller.rb]: Can you redirect to new_user_registration_path instead ?

File: app/controllers/teams_controller.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 class TeamsController < ApplicationController
     def index
         title('Team')
-        @users = User.where(role: ['administrator', 'moderator', 'developper'])
+        @users = User.where(role: ['administrator', 'moderator'])
```
Comment (eternialz) [inline app/controllers/teams_controller.rb]: Isn't it possible to use User::Roles.team here ?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -150,6 +150,34 @@ def delete
...
+            report.reportable = @user
+            report.save
+            @user.warnings << report
+            @user.save
+
+            flash[:notice] = 'User reported'
+            redirect_to user_path(@user.name)
+        else
+            flash[:error] = 'You should be connected to report someone'
+            redirect_to root_path
```
Comment (eternialz) [inline app/controllers/users_controller.rb]: Same here new_user_registration_path

File: app/models/post.rb [hunk 4]
```diff
@@ -37,13 +35,15 @@ def self.image_types
 
     validates :post_image, attached: true,
                            content_type: { in: Post.image_types },
-                           size: { less_than_or_equal_to: 50.megabytes }
+                           size: { less_than_or_equal_to: 50.megabytes },
+                           dimension: {
+                               min: ConfigHelper.min_img_width..ConfigHelper.min_img_height,
+                               max: ConfigHelper.max_img_width..ConfigHelper.max_img_height
+                           }
```
Comment (eternialz) [inline app/models/post.rb]: Nice

File: frontend/stylesheets/components/_main-menu.scss [hunk 5]
```diff
@@ -122,7 +122,8 @@ $height: 3rem;
                 z-index: $depth-2;
             }
 
-            &:hover {
+            &:hover,
+            &:focus-within {
```
Comment (eternialz) [inline frontend/stylesheets/components/_main-menu.scss]: The merge didn't work there, it should only be "&:hover {". 
"&focus-whithin" is now separated for submenu to work in Edge

File: test/services/user_service_test.rb [hunk 6]
```diff
@@ -20,14 +18,4 @@ class UserServiceTest < ActiveSupport::TestCase
...
-
-    test 'update_level_with_level_increase' do
-        @user.exp = @user.level.max_exp
-
-        assert_difference -> { @user.level.rank }, 1 do
-            UserService.update_level(@user)
-        end
-
-        assert @user.persisted?
-    end
```
Comment (eternialz) [inline test/services/user_service_test.rb]: Why ?

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (Aruisehu) [reply app/controllers/posts_controller.rb]: i didn't see the before action so it's useless for the check

File: test/services/user_service_test.rb (hunk 6 above)
Comment (Aruisehu) [reply test/services/user_service_test.rb]: Because you have no idea the merge i had to do

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (Aruisehu) [reply app/controllers/users_controller.rb]: same as posts_controller + i will redirect user who wants to report itself or a staff member to the user page he is currently visiting

**Pull request merged**